File: sdks/swift/Sources/Reactor/Internal/PublishState.swift
/// Whether this client has a sender behind a track's slot.
///
/// The session records none of this. `reactor_publish_track` is a *request* and
/// `reactor_unpublish_track` a *notification*, and neither leaves anything to
/// query — so the binding keeps it, and keeping it right is the whole job.
///
/// ## Three states, not two
///
/// A publish asked for and not yet answered is **not** published: there is no
/// sender behind the slot yet, and a frame pushed into that window is taken by
/// the FFI and dropped. Counting it as published reintroduces exactly the silent
/// failure this SDK exists to refuse; counting it as nothing tells a caller who
/// *just called* `publish()` to call `publish()`. So it is its own state, and the
/// refusal says "await the publish" rather than "publish first".
enum PublishState: Sendable {

    /// No sender behind this slot. Pushing drops the frame.
    case unpublished

    /// A publish is in flight. Still no sender: pushing still drops the frame.
    case publishing

    /// Published, and pushes reach the far end.
    case published
}

File: sdks/android/reactor-sdk-android/src/main/kotlin/inc/reactor/sdk/android/PublishState.kt
package inc.reactor.sdk.android

/**
 * Whether a sendonly slot has a sender behind it.
 *
 * Three states, not two, and the middle one is the whole point. A publish asked for and not yet
 * answered is **not** published: there is no sender behind the slot, so a push in that window is
 * taken by the FFI and dropped. Counting it as published reintroduces exactly the silent failure
 * this SDK exists to prevent; counting it as unpublished tells a caller who just called
 * `publish()` to call `publish()`.
 */
public enum class PublishState {
    /** No sender. Pushing raises. */
    UNPUBLISHED,

    /** Asked for, not yet answered. Pushing raises, and says to await the publish. */
    PUBLISHING,

    /** A sender is attached. Pushing works. */
    PUBLISHED,
}
